runtime/tr.source/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -5381,6 +5381,257 @@ bool genInstanceOfOrCheckCastNullTest(TR::Node* node, TR::CodeGenerator* cg, TR:
+ *     This funcition generates a sequence to check per site cache for object class and cast class before calling out to jitInstanceOf helper
+ */
+static
+void genInstanceOfDynamicCacheAndHelperCall(TR::Node *node, TR::CodeGenerator *cg, TR::Register *castClassReg, TR::Register *objClassReg, TR::Register *resultReg, TR_S390ScratchRegisterManager *srm, TR::LabelSymbol *doneLabel, TR::LabelSymbol *helperCallLabel, TR::LabelSymbol *dynamicCacheTestLabel, TR::LabelSymbol *branchLabel, TR::LabelSymbol *trueLabel, TR::LabelSymbol *falseLabel, bool dynamicCastClass, bool generateDynamicCache, bool cacheCastClass, bool ifInstanceOf, bool trueFallThrough )
+   {
+   TR_S390OutOfLineCodeSection *outlinedSlowPath = NULL;
+   TR::Compilation                *comp = cg->comp();
+   bool needResult = resultReg != NULL;
+   if (!castClassReg)
+      castClassReg = cg->evaluate(node->getSecondChild());
+   int32_t maxOnsiteCacheSlots = comp->getOptions()->getMaxOnsiteCacheSlotForInstanceOf();
+   TR::InstOpCode::Mnemonic loadHalfWordImmOp = TR::Compiler->target.is64Bit() ? TR::InstOpCode::LGHI : TR::InstOpCode::LHI;
+   bool needToGenerateDynamicCacheTest = generateDynamicCache && maxOnsiteCacheSlots > 0;
+   TR::Register *dynamicCacheReg = NULL;
+   int32_t addressSize = TR::Compiler->target.is64Bit() ? 8 : 4;
+   /* Layout of the writable data snippet
+    * Case - 1 : Cast class is runtime variable
+    * [UpdateIndex][ObjClassSlot-0][CastClassSlot-0]...[ObjClassSlot-N][CastClassSlot-N]
+    * Case - 2 : Cast Class is inetrface / unresolved
+    * [UpdateIndex][ObjClassSlot-0]...[ObjClassSlot-N]
+    * If there is only one cache slot, we will not have header.
+    * Last bit of cached objectClass will set to 1 indicating false cast
+    */
+   int32_t snippetSizeInBytes = ((cacheCastClass ? 2 : 1) * maxOnsiteCacheSlots * addressSize) + (addressSize * (maxOnsiteCacheSlots != 1));
+   if (needToGenerateDynamicCacheTest)
+      {
+      TR::S390WritableDataSnippet *dynamicCacheSnippet = NULL;
+      /* We can only request the snippet size of power 2, following table summarizes bytes needed for corresponding number of cache slots
+       * Case 1 : Cast class is runtime variable
+       * Case 2 : Cast class is interface / unresolved
+       * Number Of Slots |  Bytes needed for Case 1 | Bytes needed for Case 2
+       *        1        |              16          |           8
+       *        2        |              64          |           32
+       *        3        |              64          |           32
+       *        4        |              128         |           64
+       *        5        |              128         |           64
+       *        6        |              128         |           64
+       */
+      int32_t requestedBytes = 1 << (int) (log2(snippetSizeInBytes-1)+1);
+      traceMsg(comp, "Requested Bytes = %d\n",requestedBytes);
+      uintptr_t *initialSnippet = new (cg->trHeapMemory()) uintptr_t[requestedBytes/sizeof(uintptr_t)];
```

Comment:
The `CreateConstant` API copies the bytes you pass to it into it's internal buffer. This means `initialSnippet ` can be stack allocated instead of heap allocated. Are we able to save ourselves some memory here by stack allocating this array on the function frame?

Reply:
I had `initialSnippet` heap allocated as the size of requested bytes can be changed. We can certainly have this stack allocated by using a large array (Maximum size of Snippet we are going to allocate is 128 bytes) and we can use constant array here.

Reply:
Let's stack allocate a reasonably sized array and add an assert to make sure `getMaxOnsiteCacheSlotForInstanceOf` doesn't overflow the array.